rsocket-core/src/main/java/io/rsocket/util/ReconnectingRSocket.java

```diff
@@ -0,0 +1,539 @@
+        Predicate<Throwable> errorPredicate) {
+      this.parent = parent;
+      this.mapper = mapper;
+      this.errorPredicate = errorPredicate;
+    }
+
+    @Override
+    public void subscribe(CoreSubscriber<? super T> actual) {
+      if (state == NONE && STATE.compareAndSet(this, NONE, SUBSCRIBED)) {
+        this.actual = actual;
+        parent.subscribe(this);
+      } else {
+        Operators.error(actual, new IllegalStateException("Only a single Subscriber allowed"));
+      }
+    }
+
+    @Override
+    public void accept(RSocket rSocket) {
+      if (rSocket == null) {
+        Operators.error(actual, new CancellationException("Disposed"));
+      }
+
+      Flux<? extends T> source;
+      try {
+        source = this.mapper.apply(rSocket);
+        source.subscribe(this);
+      } catch (Throwable e) {
+        Exceptions.throwIfFatal(e);
+        Operators.error(actual, e);
+      }
+    }
+
+    @Override
+    public Context currentContext() {
+      return actual.currentContext();
+    }
+
+    @Nullable
+    @Override
+    public Object scanUnsafe(Attr key) {
+      if (key == Attr.PARENT) return s;
+      if (key == Attr.ACTUAL) return parent;
+      if (key == Attr.TERMINATED) return done;
+      if (key == Attr.CANCELLED) return s == Operators.cancelledSubscription();
+
+      return null;
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+      this.s = s;
+      actual.onSubscribe(this);
+    }
+
+    @Override
+    public void onNext(T payload) {
+      actual.onNext(payload);
+    }
+
+    @Override
+    public void onError(Throwable t) {
+      if (done) {
+        Operators.onErrorDropped(t, actual.currentContext());
+        return;
+      }
+
+      final CoreSubscriber<? super T> actual = this.actual;
+
+      if (errorPredicate.test(t)) {
+        this.actual = null;
+        STATE.compareAndSet(this, SUBSCRIBED, NONE);
+      } else {
+        done = true;
+      }
+      actual.onError(t);
+    }
+
+    @Override
+    public void onComplete() {
+      if (done) {
+        return;
+      }
+      done = true;
+      actual.onComplete();
+    }
+
+    @Override
+    public void request(long n) {
+      s.request(n);
+    }
+
+    public void cancel() {
+      if (STATE.getAndSet(this, CANCELLED) != CANCELLED) {
+        s.cancel();
+      }
+    }
+  }
+
+  public static class Builder {
+
+    private static final Predicate<Throwable> DEFAULT_ERROR_PREDICATE =
+        throwable ->
+            throwable instanceof ClosedChannelException
+                || throwable instanceof ConnectionCloseException
+                || throwable instanceof ConnectionErrorException;
+
+    Duration backoffMin;
+    Duration backoffMax;
+    Supplier<Mono<RSocket>> sourceSupplier;
+    Scheduler scheduler = Schedulers.parallel();
+    Predicate<Throwable> errorPredicate = DEFAULT_ERROR_PREDICATE;
+
+    public Builder withSourceRSocket(Supplier<Mono<RSocket>> sourceSupplier) {
```

Comment:
why not just Mono<RSocket> here?  That could be an instance of Mono.defer called externally.  I always assumed Mono.defer taking Supplier was more about having a SAM and being able to call some sync block that returned e.g. Mono.just(...)